File: Question2.java
import java.io.*;
import java.util.*;


public class Question2 {

  private String[] arr;

  public String[] getArr(){
    return arr;
  }

  Question2(int n){
    arr = new String[n];
    Arrays.fill(arr, "closed");
  }

  public void checkDoors (int times){
        for (int i=1; i <= times; i++)
        for (int d=1; d <= arr.length; d++)
            if (d % i == 0)
                toggleDoor(d-1);
  }

  private void toggleDoor (int door) {
    if (arr[door] == "open")
      arr[door] = "closed";
    else
      arr[door] = "open";
  }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        System.out.print("Enter number of doors: ");
        int n = in.nextInt();
        Question2 q = new Question2(n);
        System.out.print("How many passes: ");
        int p = in.nextInt();
        q.checkDoors(p);
        String[] arr = q.getArr();
        for(int i=1; i<=n; i++){
          System.out.println(String.format("%s - %s",i,arr[i-1]));
        }
        in.close();
    }
}

File: question2.rb
#n doors in a row are all initially closed.
#You make n passes by the doors.
#The first time through, you visit every door and toggle the door (if the door is closed, you open it; if it is open, you close it).
#The second time you only visit every 2nd door (door #2, #4, #6, ...).
#The third time, every 3rd door (door #3, #6, #9, ...), etc, until you only visit the nth door.

#Question: What state are the doors in after the last pass? Which are open, which are closed?
#
class Question2

  attr_reader :arr

  def initialize (n)
    @arr = Array.new(n,'closed')
  end

  def check_doors (n)
    for i in (1..n) do
      for d in (1..@arr.length)
        if d%i == 0  #visit the door
          toggle_door(d)
        end
      end
    end
  end

private

  def toggle_door (d)
    if @arr[d-1] == 'open'
      @arr[d-1] = 'closed'
    else
      @arr[d-1] = 'open'
    end
  end
end
